lib/new_relixir/statman.ex

```diff
@@ -0,0 +1,186 @@
+        else
+          []
+        end
+      _ -> []
+    end
+  end
+
+  def summary(values) do
+    summary(values, {0, 0, 0, 0, 0})
+  end
+  def summary([], {count, sum, min, max, sum2}) do
+    [
+      count,
+      sum / 1_000_000,
+      sum / 1_000_000,
+      min / 1_000_000,
+      max / 1_000_000,
+      sum2 / 1_000_000_000
+    ]
+  end
+  def summary([v|rest], {count, sum, min, max, sum2}) do
+    summary(rest, {count+1, sum+v, v < min && v || min, v > max && v || max, sum2+v*v})
+  end
+  def transform_histogram({key, values}) do
+    data = summary(values)
+
+    case key do
+      {scope, {:db, segment}} when is_binary(scope) ->
+        [
+         [%{name: "Database/#{to_bin(segment)}", scope: scope2bin(scope)}, data],
+         [%{name: "Database/allWeb", scope: ""}, data],
+         [%{name: "Database/all", scope: ""}, data]
+        ]
+
+      {scope, {:ext, host}} when is_binary(scope) and is_binary(host) ->
+        [
+         [%{name: "External/all", scope: ""}, data],
+         [%{name: "External/allWeb", scope: ""}, data],
+         [%{name: "External/#{host}", scope: ""}, data],
+         [%{name: "External/#{host}/all", scope: ""}, data],
+         [%{name: "External/#{host}", scope: scope2bin(scope)}, data]
+        ]
+
+      {{:background, scope}, :total} when is_binary(scope) ->
+          [[%{name: bgscope2bin(scope), scope: ""}, data]]
+
+      {{:background, scope}, {class, segment}} when is_binary(scope) ->
+          [[%{name: class2bin(class) <> "/" <> to_bin(segment), scope: bgscope2bin(scope)}, data]]
+
+      {scope, {class, segment}} when is_binary(scope) ->
+          [[%{name: class2bin(class) <> "/" <> to_bin(segment), scope: scope2bin(scope)}, data]]
+
+      {scope, :total} when is_binary(scope) ->
+          [[%{name: "WebTransaction/Uri/#{scope}", scope: ""}, data]]
+
+      {a, b} when is_atom(a) and is_atom(b) ->
+          [[%{name: "OtherTransaction/#{a}/#{b}", scope: ""}, data]]
+      _ ->
+          []
+    end
+  end
+
+  def webtransaction_total(ms) do
+    name = "WebTransaction"
+    n    = Enum.sum(pluck(name, 1, ms))
+    sum  = Enum.sum(pluck(name, 2, ms))
+    min  = Enum.min(pluck(name, 4, ms))
+    max  = Enum.max(pluck(name, 5, ms))
+    sum2 = Enum.sum(pluck(name, 6, ms))
+
+    [%{name: "HttpDispatcher", scope: ""}, [n, sum, sum, min, max, sum2]]
+  end
+
+  def db_total(ms) do
+    name = "Database"
+    n    = Enum.sum(pluck(name, 1, ms))
+    sum  = Enum.sum(pluck(name, 2, ms))
+    min  = Enum.min(pluck(name, 4, ms))
+    max  = Enum.max(pluck(name, 5, ms))
+    sum2 = Enum.sum(pluck(name, 6, ms))
+
+    [%{name: "Database/all", scope: ""}, [n, sum, sum, min, max, sum2]]
+  end
+
+  def errors_total(errors) do
+    data = [length(errors), 0.0, 0.0, 0.0, 0.0, 0.0]
+    [[%{name: "Errors/all", scope: ""}, data],
+     [%{name: "Errors/allWeb", scope: ""}, data],
+     [%{name: "Instance/Reporting", scope: ""}, data]
+    ]
+  end
+
+  def pluck(_, _, []), do: [0]
```

Comment:
The logic here is a little hard to follow, I'd recommend refactoring it:

```elixir
# Before
def pluck(_, _, []), do: [0]
def pluck(name, n, list) do
  Enum.map(list, fn(elem) ->
    if is_list(elem) && is_list(hd(elem)) do
      pluck(name, n, elem)
    else
      get_nth(elem, name, n-1)
    end
  end)
  |> List.flatten
end

defp get_nth([[_, []]], _, _), do: 0
defp get_nth([_, []], _, _), do: 0
defp get_nth([[struct, data_list]], name, n), do: get_nth([struct, data_list], name, n)
defp get_nth([struct, data_list], name, n) do
  if String.contains?(struct[:name], name) do
    if struct[:scope] == "" do
      Enum.at(data_list, n) || 0
    else
      0
    end
  else
    0
  end
end

# After
def pluck(type, position, metrics) do
  metrics
  |> Enum.filter(fn [%{name: name}, _data] -> String.contains?(name, type) end)
  |> Enum.map(fn [_metric, data] -> Enum.at(data, position) || 0 end)
end

# Then the calls to `min` and `max` need to provide a default value:

# Before
min  = Enum.min(pluck(name, 4, ms))
max  = Enum.max(pluck(name, 5, ms))

# After
min  = Enum.min(pluck(name, 4, ms), fn -> 0 end)
max  = Enum.max(pluck(name, 5, ms), fn -> 0 end)
```

Reply:
Done in 63f7bff23e07f70e761be5202e1cfe58b4092f4d

Reply:
Nice! 👌